[ACTION] Support Sendgrid Actions

sergioeliot2039 opened the pull request:
This is a new PR for Sendgrid Actions. The previous [one](https://github.com/PipedreamHQ/pipedream/issues/1248) had modified unintendedly Reddit and Mailgun files.

Is there a specific app this action is for?
Sendgrid

Add Emails to Global Suppression | Adds recipient email address to the global suppression group.
Add or Update a Contact | Creates new contact or updates existing based on the email address.
Add or Update a Recipient (Legacy) | Adds a new recipient or updates an existing one based on the email address.
Add Recipients to a List (Legacy) | Adds existing recipients to a list.
Create a List | Creates a new list.
Delete a Global Suppression | Removes an email address from the global suppressions group.
Delete a List | Deletes a list.
Delete all Bounces | Deletes all bounces.
Delete Blocks | Deletes either all or specific emails on your block list.
Delete Bounces | Deletes specified bounces.
Delete Contacts | Deletes one or more contacts.
Delete Recipient from a List (Legacy) | Deletes recipient from a specific list.
Delete Recipients (Legacy) | Deletes one or more recipients.
Get a Block | Retrieves a specific email address from your block list.
Get a Global Suppression | Retrieves a specific global suppression email.
Make an API Call | Performs an arbitrary authorized API call.
Remove Contacts from a List | Removes contacts from the given list. The contacts will not be deleted.
Send an Email | Sends an email.
Validate an Email | Validates an email address.
Get all Bounces | Retrieves all of your bounces.
Get all Contacts | Retrieves all your contacts.
Get all Lists | Retrieves all your contact lists.
Get all Recipients | Retrieves all of your recipients or recipients under a specified list.
List Blocks | Lists all email addresses that are currently on your block list.
List Global Suppressions | Lists all email addresses that are globally suppressed.
Search Contacts | Searches for contacts.

Please provide a link to the relevant API docs for the specific service / operation.
https://sendgrid.com/docs/api-reference/
For Legacy endpoints: https://www.twilio.com/docs/sendgrid/api/v2/using_the_web_api

jverce commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -108,5 +114,212 @@ module.exports = {
+        })
+      );
+    },
+    async getAllContactLists(pageSize, pageToken) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          path: `/marketing/lists`,
+          params: {
+            page_size: pageSize,
+            page_token: pageToken,
+          },
+        })
+      );
+    },
```
As mentioned above, `pageSize` and `pageToken` are API details and should not be exposed by this method.
Instead, we should accept a `maxItems` argument to use as an upper bound on the collection of retrieved records.

```suggestion
    async *getAllContactLists(maxItems) {
      let url = `${this._apiUrl()}/marketing/lists`;

      while (url && maxItems > 0) {    
        const params = {
          page_size: Math.min(maxItems, 1000),
        };
        const requestConfig = {
          ...this._makeRequestConfig(),
          params,
        };
        const { data } = this._withRetries(
          () => axios.get(url, requestConfig),
        );

        const contactLists = data.result.slice(0, maxItems);
        for (const contactList of contactLists) {
          yield contactList
        }

        maxItems -= contactLists.length;
        url = data._metadata.next;
      }
    },
```

sergioeliot2039 replied:
Understood. Will be using similar approach on other actions related to an API endpoint having pageSize/pageToken..

jverce commented on components/sendgrid/actions/get-all-contacts/get-all-contacts.js:
```diff
@@ -0,0 +1,26 @@
+  key: "sendgrid-get-all-contacts",
+  name: "Get All Contacts",
+  description: ".",
+  version: "0.0.4",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/get-all-recipients/get-all-recipients.js:
```diff
@@ -0,0 +1,28 @@
+  key: "sendgrid-get-all-recipients",
+  name: "Get All Recipients",
+  description: "Allows you to get all of your Marketing Campaigns recipients.",
+  version: "0.0.5",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/get-all-recipients/get-all-recipients.js:
```diff
@@ -0,0 +1,28 @@
+  type: "action",
+  props: {
+    sendgrid,
+    page: {
+      type: "integer",
+      label: "Page",
+      description: "Page index of first recipients to return (must be a positive integer)",
+      optional: true,
+    },
+    pageSize: {
+      type: "integer",
+      label: "Page Size",
+      description:
+        "Number of recipients to return at a time (must be a positive integer between 1 and 1000)",
+      optional: true,
+    }
```
Same comments as above

jverce commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,40 @@
+      description: "The end of the time range when a blocked email was created (inclusive). This is a unix timestamp.",
+      optional: true,
+    },
+    limit: {
+      type: "integer",
+      label: "Limit",
+      description:
+        "Limit the number of results to be displayed per page.",
+      optional: true,
+    },
+    offset: {
+      type: "integer",
+      label: "Offset",
+      description: "The point in the list to begin displaying results.",
+      optional: true,
+    }
```
Similar to my comments above regarding pagination, these props should not be exposed to the user since they're out of their control. We should instead accept a parameter like _Number of blocked emails to return_:
![image](https://user-images.githubusercontent.com/5479513/121625998-6c0a0d00-ca29-11eb-8741-6cf66bf3d50e.png)

jverce commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,40 @@
+  key: "sendgrid-list-global-supressions",
+  name: "List Global Supression",
+  description: "Allows you to get a list of all email address that are globally suppressed.",
+  version: "0.0.4",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,40 @@
+  key: "sendgrid-list-blocks",
+  name: "List Blocks",
+  description: "Allows you to list all email addresses that are currently on your blocks list.",
+  version: "0.0.6",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,40 @@
+      description: "Refers end of the time range in unix timestamp when an unsubscribe email was created (inclusive).",
+      optional: true,
+    },
+    limit: {
+      type: "integer",
+      label: "Limit",
+      description:
+        "Limit the number of results to be displayed per page.",
+      optional: true,
+    },
+    offset: {
+      type: "integer",
+      label: "Offset",
+      description: "The point in the list of results to begin displaying global suppressions.",
+      optional: true,
+    }
```
Same as above.

jverce commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,39 @@
+  key: "sendgrid-remove-a-contact-from-a-list",
+  name: "Remove A Contact From A List",
+  description: "Allows you to remove contacts from a given list.",
+  version: "0.0.3",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/search-contacts/search-contacts.js:
```diff
@@ -0,0 +1,26 @@
+  key: "sendgrid-search-contacts",
+  name: "Search Contacts",
+  description: "Searches contacts with a SGQL query.",
+  version: "0.0.4",
```
```suggestion
  version: "0.0.1",
```

sergioeliot2039 replied:
Well, since I am using `pd publish` for testing out my actions, I think Pipedream`s publishing mechanism would fail if we left version `0.0.1`, or lower than the great version used. In a previous review @dylburger suggested, after the PR was approved, to publish components using `0.1.0`. So while I do is as I commit to the repo I will use `0.0.1` but set to `0.10` for merging,. after the PR is approved.

dylburger replied:
You should set the version in the merged PR to `0.0.1` in this case. The version tied to your published component in your account is distinct from the version that will be used to publish, so even if you're on `0.0.4` in your account, you can publish `0.0.1` to the global registry.

sergioeliot2039 replied:
Gotcha! Will do then.

jverce commented on components/sendgrid/actions/send-email/send-email.js:
```diff
@@ -0,0 +1,173 @@
+
+module.exports = {
+  key: "sendgrid-send-an-email",
+  name: "Send An Email",
```
```suggestion
  name: "Send an Email",
```

jverce commented on components/sendgrid/actions/send-email/send-email.js:
```diff
@@ -0,0 +1,173 @@
+module.exports = {
+  key: "sendgrid-send-an-email",
+  name: "Send An Email",
+  description: "Sends an email.",
```
```suggestion
  description: "This action sends a personalized e-mail to the specified recipients",
```

jverce commented on components/sendgrid/actions/send-email/send-email.js:
```diff
@@ -0,0 +1,173 @@
+  key: "sendgrid-send-an-email",
+  name: "Send An Email",
+  description: "Sends an email.",
+  version: "0.0.35",
```
```suggestion
  version: "0.0.1",
```

jverce commented on components/sendgrid/actions/send-email/send-email.js:
```diff
@@ -0,0 +1,173 @@
+    },
+  },
+  async run() {
+    if (
+      !this.personalizations ||
+      !this.fromEmail ||
+      !this.subject ||
+      !this.content
+    ) {
+      throw new Error(
+        "Must provide personalizations, fromEmail, subject, and content parameters."
+      );
+    }
+    const personalizations = JSON.parse(this.personalizations);
+    const content = JSON.parse(this.content);
+    const attachments = this.attachments ? JSON.parse(this.attachments) : null;
+    const categories = this.categories ? JSON.parse(this.categories) : null;
+    const replyTo = this.replyToEmail
+      ? { email: this.replyToEmail, name: this.replyToName }
+      : null;
```
Validation is painful, so perhaps we could use [`validate.js`](http://validatejs.org/) for all these actions?

For example:
```js
const validate = require("validate.js");

module.exports = {
  run() {
    const fromEmailValidation = validate.single(this.fromEmail, {
      presence: true,
      email: true,
    });
    if (fromEmailValidation) {
      throw new Error(fromEmailValidation);
    }
  },
};
```

sergioeliot2039 replied:
Sure, I've been using `validate.js` since all these changes were requested.

sergioeliot2039 commented on components/sendgrid/actions/add-email-to-global-supression/add-email-to-global-supression.js:
```diff
@@ -0,0 +1,27 @@
+      throw new Error("Must provide recipientEmails parameter.");
+    }
+    const recipientEmails = JSON.parse(this.recipientEmails);
+    return await this.sendgrid.addEmailToGlobalSupression(recipientEmails);
```
Added a validation to check `Recipient Emails` parameter is an Array.

sergioeliot2039 commented on components/sendgrid/actions/add-or-update-contact/add-or-update-contact.js:
```diff
@@ -0,0 +1,36 @@
+      throw new Error("Must provide contacts parameter.");
+    }
+    const listIds = this.listIds ? JSON.parse(this.listIds) : null;
+    const contacts = JSON.parse(this.contacts);
```
Added checks on Contacts parameter.

sergioeliot2039 commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,27 @@
+    },
+  },
+  async run() {
+    return await this.sendgrid.deleteContacts(this.deleteAllContacts, this.ids);
```
No, because specifiedly, both parameters are optional.
However, the user must provide one or the other.
`You must set either ids or delete_all_contacts.`
I changed the code force the `deleteAllContacts` prop to be boolean.

sergioeliot2039 commented on components/sendgrid/actions/delete-global-supression/delete-global-supression.js:
```diff
@@ -0,0 +1,31 @@
+      label: "Email",
+      description:
+        "The email address you want to remove from the global suppressions group.",
+      useQuery: true,
+      async options() {
+        const options = [];
+        const globalSupressions = await this.sendgrid.listGlobalSupressions();
+        for (const globalSupression of globalSupressions) {
+          options.push(globalSupression.email);
+        }
+        return options;
+      },
```
Great catch, set the prop type so that the value can be referenced to a previous step in the workflow.

sergioeliot2039 commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -27,13 +27,23 @@ module.exports = {
         headers,
       };
     },
+    async _makeRequest(opts) {
+      if (!opts.headers) opts.headers = {};
+      opts.headers.authorization = `Bearer ${this._authToken()}`;
+      opts.headers["user-agent"] = "@PipedreamHQ/pipedream v0.1";
+      opts.headers["Content-Type"] = "application/json";
+      const { path } = opts;
+      delete opts.path;
+      opts.url = `${this._apiUrl()}${path[0] === "/" ? "" : "/"}${path}`;
+      return (await axios(opts)).data;
+    },
```
Ok, I remade most all of the requests method to use `_makeRequestConfig` but there is one problematic, will discuss on Slack.

sergioeliot2039 commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -108,5 +114,212 @@ module.exports = {
     async disableSignedWebhook() {
       return this._setSignedWebhook(false);
     },
+    async addEmailToGlobalSupression(recipientEmails) {
+      return await this._withRetries(() =>
+        this._makeRequest({
+          method: "POST",
+          path: `/asm/suppressions/global`,
+          data: {
+            recipient_emails: recipientEmails,
+          },
+        })
+      );
+    },
```
This was done.

sergioeliot2039 commented on components/sendgrid/actions/get-a-global-supression/get-a-global-supression.js:
```diff
@@ -0,0 +1,32 @@
+      type: "string",
+      label: "Email",
+      description: "The email address of the global suppression you want to retrieve.",
+      useQuery: true,
+      async options() {
+        const options = [];
+        const globalSupressions = await this.sendgrid.listGlobalSupressions();
+        for (const globalSupression of globalSupressions) {
+          options.push(globalSupression.email);
+        }
+        return options;
+      },
```
Change made.

jverce commented on components/sendgrid/actions/get-all-bounces/get-all-bounces.js:
```diff
@@ -0,0 +1,29 @@
+  type: "action",
+  props: {
+    sendgrid,
+    startTime: {
+      type: "integer",
+      label: "Start Time",
+      description:
+        "Refers start of the time range in unix timestamp when a bounce was created (inclusive).",
+      optional: true,
+    },
+    endTime: {
+      type: "integer",
+      label: "End Time",
+      description:
+        "Refers end of the time range in unix timestamp when a bounce was created (inclusive).",
+      optional: true,
+    },
```
Why are we not validating these? Are we relying on the Sendgrid API call to do it for us?
Perhaps we could do some basic validation at least, for example make sure that:
- Both numbers are greater than 0
- `startTime` is greater than `endTime`

sergioeliot2039 replied:
Ok, adding some basic validation.

jverce commented on components/sendgrid/actions/get-contact-lists/get-contact-lists.js:
```diff
@@ -0,0 +1,44 @@
+  async run() {
+    const constraints = {
+      numberOfLists: {
+        presence: true,
```
We should also verify that this number is positive, right? At least so that the user gets a proper error message if they provide the wrong value.

sergioeliot2039 replied:
I added a validation method for positive ranges.

jverce commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,55 @@
+      },
+      contactIds: {
+        presence: true,
+      },
```
```suggestion
        type: "array",
      },
```

jverce commented on components/sendgrid/actions/search-contacts/search-contacts.js:
```diff
@@ -0,0 +1,31 @@
+    },
+  },
+  async run() {
+    const validate = require("validate.js");
```
Can you move this to the top, like the other `require`s?

sergioeliot2039 replied:
Sure, I think I already started to move the `require`s to the top in other action files.

jverce commented:
@sergioeliot2039 there's a bunch of linting and spellcheck errors that must be addressed before we can merge this PR.
To fix the linting issues, you can run the following commands at the root of this repo:
```shell
npm install
npx eslint --fix components/sendgrid/actions
```

jverce commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -27,13 +59,23 @@ module.exports = {
       const authToken = this._authToken();
       const headers = {
-        "Authorization": `Bearer ${authToken}`,
+        Authorization: `Bearer ${authToken}`,
```
```suggestion
        "Authorization": `Bearer ${authToken}`,
        "Content-Type": "application/json",
```

jverce commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -27,13 +59,23 @@ module.exports = {
         headers,
       };
     },
+    async _makeRequest(opts) {
+      if (!opts.headers) opts.headers = {};
+      opts.headers.authorization = `Bearer ${this._authToken()}`;
+      opts.headers["user-agent"] = "@PipedreamHQ/pipedream v0.1";
+      opts.headers["Content-Type"] = "application/json";
```
```suggestion
    async _makeRequest(opts = {}) {
      opts = {
        ...this._makeRequestConfig(),
        ...opts,
      };
```

jverce reviewed (changes requested):
Related to our conversation in Slack

jverce commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,56 @@
+      optional: true,
+    },
+    ids: {
+      type: "object",
+      label: "Ids",
+      description: "An array of contact IDs to delete.",
+      optional: true,
```
An object here makes no sense IMO
![image](https://user-images.githubusercontent.com/5479513/126414537-da33a08d-246d-499e-998f-752bf490b197.png)

The way to inject values into that prop is by setting its type to `string` (as discussed offline in Slack), and do the runtime validation during the action execution to make sure that the injected value complies with the expected type/format. In other words, making this a prop of type `string` allows users to specify a reference to another value (which will then be injected by the platform) and it is this last value the one that we must validate.

```suggestion
      type: "string",
      label: "Contact ID's",
      description: "An array of contact IDs to delete",
      optional: true,
```

sergioeliot2039 replied:
Ok, I had used object instead of string when receiving arrays to that we could leverage the validation capabilities from the `validate.js` library. Having the string we will have to parse and maybe use custom validator under `validate.js`

jverce commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,56 @@
+const common = require("../common");
+
+module.exports = {
+  key: "sendgrid-delete-contacts",
```
```suggestion
  ...common,
  key: "sendgrid-delete-contacts",
```

jverce commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,56 @@
+  version: "0.0.1",
+  type: "action",
+  props: {
+    sendgrid,
```
```suggestion
    ...common.props,
```

jverce commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,56 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common,
+  },
```
When extending another component, make sure you refer to the correct nested attribute
```suggestion
  methods: {
    ...common.methods,
  },
```

However, in this case it shouldn't be necessary if you already do this at the beginning of the module definition:
```js
module.exports = {
  // This essentially sets every attribute defined in `common`
  // to be also defined in the current module
  ...common,
  props: {
    // Since re-defining the `props` attribute overrides the one previously
    // defined by `common`, we must explicitly say that we want those
    // as well.
    ...common.props,
    newProp: {},
  },
};
```

sergioeliot2039 commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,56 @@
+      type: "boolean",
+      label: "Delete All Contacts?",
+      description: "This parameter allows you to delete all of your contacts. This can not be used with the `Ids` parameter.",
+      optional: true,
```
All right, thanks for providing criteria for when to use `default` values. I had added default values in delete-blocks and delete-bounces so not sure why I missed adding defaults in this one.

compwright commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -12,10 +12,42 @@ module.exports = {
     },
+    _emailValidationsUrl() {
+      const baseUrl = this._apiUrl();
+      return `${baseUrl}/validatiogns/email`;
```
Is this a typo? `/validatiogns/`

compwright commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -12,10 +12,42 @@ module.exports = {
     _apiUrl() {
```
I'm really against making methods that do nothing except return an API URL. Can we get rid of these?

They clutter up the app file with a bunch of unnecessary methods. The API URLs aren't going to change that often, and it's easier to identify the API being called if you have the entire URL where the call is being made. I see no net benefit to having them here.

sergioeliot2039 replied:
Noted. I got rid of the methods returning API URL for Sendgrid actions only. For methods related to Sendgrid sources, they were left unchanged.

compwright commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -28,14 +60,30 @@ module.exports = {
         headers,
       };
     },
+    async _makeRequest(opts = {}) {
```
Sendgrid has a well-maintained official API client. Can we use [sendgrid-nodejs](https://github.com/sendgrid/sendgrid-nodejs) instead of Axios?

sergioeliot2039 replied:
All Sendgrid action components were migrated to use sendgrid-nodejs instead of Axios.
Sendgrid sources, which aren't related to my assignemnt, weren't changed.

compwright reviewed (changes requested):
Will re-review when these are addressed.

sergioeliot2039 commented:
Update: I have completed migration of Sendgrid components to use the Sendgrid SDK. I will next review the other change requests. Should be ready for another review soon.

sergioeliot2039 commented:
Hi @dylburger 
cc @compwright 

Some of the requested changes I am tackling for sendgrid actions, may impact the overall sendgrid components.
1. Using Sendgrid SDK. do you want Sendgrid sources to be migrated to use the SDK?
2. Getting rid of methods that return an URL. Should we remove methods returning an URL and change Sendgrid sources accordingly?

I wanted to check before I go ahead with this,  since the issue I am tackling / have assigned is for Sendgrid actions. I am unsureif you'd want to change Sendgrid Sources now, or mix code using axios (for Sources) and the SDK (for Actions).

sergioeliot2039 commented:
@compwright Support Sengrid actions is ready for  a new review.

Note that Sendgrid sources were changed only for ESLint fixups. a per this discussion on [Slack](https://pipedream-users.slack.com/archives/C01E5KCTR16/p1636478530040600)

jcortes commented on components/sendgrid/actions/add-email-to-global-supression/add-email-to-global-supression.js:
```diff
@@ -0,0 +1,23 @@
+    },
+  },
+  async run() {
+    return await this.sendgrid.addEmailToGlobalSupression(this.recipientEmails);
```
As far as I know the `await` keyword here is not necessary. Since you are not doing anything with a `try/catch` block. So you can remove the `await` and anywhere else you are using it along with `return` and without using `try/catch`

sergioeliot2039 replied:
oh yeah I remember we discussed this for Shortcut (formerly Clubhouselio)

> Hi [@sergioeliot2039](https://github.com/sergioeliot2039) I've found [this article](https://jakearchibald.com/2017/await-vs-return-vs-return-await/) very interesting in order to totally understand await vs return vs return await. I hope it helps!

> In [components/shortcut/actions/create-story/create-story.js](https://github.com/PipedreamHQ/pipedream/pull/1383#discussion_r762045467):

For sure, I'll get to this

sergioeliot2039 replied:
I made this change as requested.

jcortes replied:
I still see it in the code 🤔

sergioeliot2039 replied:
Right, looks like I oversaw a few of those. But most had been addressed, All of them should be good now.

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -1,6 +1,7 @@
 const axios = require("axios");
```
Shouldn't we be migrating from `axios` lib to `@pipedream/platform` according to [this doc](https://pipedream.com/docs/pipedream-axios/#pipedream-platform-axios)?

sergioeliot2039 replied:
Hi @dannyroosevelt @dylburger 
just checking in to see if you want this change actually implemented? when i started coding I remember lib `@pipedream/platform` was meant to be use on the Pipedream UI workflow builder. Now I see it is documented for component actions as well in the documentation, and this is enough reason to doing the change commented by @jcortes 

Now, I understand because of the upcoming release of the new builder, we want to be faster in publishing compoents.
In view of this, do you want this change 

> axios lib to @pipedream/platform
for Sendgrid components?

sergioeliot2039 replied:
As per this [Slack dicussion](https://pipedream-users.slack.com/archives/G01J50YGJGP/p1645187574309249)
this and other changes noted in [this comment](https://github.com/PipedreamHQ/pipedream/pull/1288#issuecomment-1049088143), I won't be addressing them because they pertain to Sendgrid sources.

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -12,9 +13,17 @@ module.exports = {
     _apiUrl() {
       return "https://api.sendgrid.com/v3";
```
As a preference I would suggest having these kind of values in constants. Something like
```javascript
const BASE_URL = "https://api.sendgrid.com";
const VERSION_PATH = "/v3";
```
in that way the maintenance of these would be straightforward in case the api decides to upgrade versions

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -77,39 +86,25 @@ module.exports = {
-      return this._withRetries(
-        () => axios.get(url, requestConfig),
-      );
+      return this._withRetries(() => axios.get(url, requestConfig));
```
I would also go with the only function that can deal with `axios` calls, in this case it might be called `_makeRequest(customConfig)` in that way the `axios` lib is not being called in other functions like this one but abstracted.

sergioeliot2039 replied:
hey @jcortes regarding `axios`... i checked the code base, and `axios` lib is used on sources, which are already  published and were authored by @jverce 
this pr is about sendgrid actions, not source. I think we shouldn't modify this code. don't you think?
thoughts? @dylburger @dannyroosevelt

sergioeliot2039 replied:
As per this [Slack dicussion](https://pipedream-users.slack.com/archives/G01J50YGJGP/p1645187574309249)
this and other changes noted in https://github.com/PipedreamHQ/pipedream/pull/1288#issuecomment-1049088143, I won't be addressing them because they pertain to Sendgrid sources.

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+    async addEmailToGlobalSupression(recipientEmails) {
+      const config = {
+        method: "POST",
+        url: "/v3/asm/suppressions/global",
```
If you create a function such as `_makeRequest(customConfig)` you would just need to pass parameters like this:
```javascript
const customConfig = {
  method: "post",
  path: "/asm/suppressions/global",
  ...args,
};
await this._makeApiRequest(customConfig);
```

sergioeliot2039 replied:
I just addressed this comment.
but named it `_makeClientRequest`.

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+          break;
+        }
+        url = data._metadata.next.replace("https://api.sendgrid.com", "");
+      } while (true && contactLists.length < maxItems);
```
Is this `true` boolean really needed here?

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+        }
+        const links = data[0].headers.link.split(",");
+        url = "";
+        links.forEach( (link) => {
```
```suggestion
        links.forEach((link) => {
```

jcortes replied:
Fix this one too

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+        links.forEach( (link) => {
+          let next = "";
+          let last = "";
+          if ( link.indexOf("next") > -1) {
```
```suggestion
          if (link.indexOf("next") > -1) {
```

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+              .replace(">", "");
+            url = next;
+          }
+          if ( link.indexOf("last") > -1) {
```
```suggestion
          if (link.indexOf("last") > -1) {
```

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+            last = link.substring(0, idx).replace("<", "")
+              .replace(">", "");
+          }
+          const nextArr = [];
+          nextArr.push(next);
+          if (nextArr.includes(last)) {
+            lastIteration = true;
+          }
```
it seems like you just need something like
```suggestion
          lastIteration = next === last;
```

jcortes commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -118,5 +113,429 @@ module.exports = {
+     * the recipient was added to the global suppression list, `email` for the email address of
+     * the recipient who is globally suppressed.
+     */
+    async listGlobalSupressions(startTime, endTime, maxItems) {
```
It seems like this function along with `listBlocks` function have similar patterns. How about abstracting this and put them in a single function call?

sergioeliot2039 replied:
Yeah that should do. Abstracting out into a single function call was easy.

jcortes commented on components/sendgrid/actions/add-email-to-global-supression/add-email-to-global-supression.js:
```diff
@@ -0,0 +1,23 @@
+  ...common,
+  key: "sendgrid-add-email-to-global-supression",
+  name: "Add Email To Global Supression",
+  description:
+    "Allows you to add one or more email addresses to the global suppressions group.",
```
Can you also point out in the `description` property the api endpoint docs on each action/source?  something like:
```suggestion
  description:
    "Allows you to add one or more email addresses to the global suppressions group. [See the docs here](https://sendgrid.api-docs.io/v3.0/suppressions-global-suppressions/add-recipient-addresses-to-the-global-suppression-group)",
```

jcortes commented on components/sendgrid/actions/common.js:
```diff
@@ -0,0 +1,89 @@
+      };
+    },
+    /**
+     * Checks if an object is an array or an JSON array string, returns an error message when the
```
```suggestion
     * Checks if an object is an array or a JSON array string, returns an error message when it
```

jcortes commented on components/sendgrid/actions/common.js:
```diff
@@ -0,0 +1,89 @@
+        throw new Error(errorMsg);
+      }
+    },
+    /**
+     * Returns `undefined` when `value` is an empty string or `null`.
+     *
+     * @param {object} value the value to check for returning `undefined`.
+     * @returns If `value` is defined, it will return `value`. Otherwise on an empty string, or
+     * `null` it will return `undefined`.
+     */
+    convertEmptyStringToUndefined(value) {
+      return value || undefined;
+    },
```
If value is the boolean `false` or the number `0` it will also return `undefined`

sergioeliot2039 replied:
Eagle eyed catch))

jcortes reviewed (changes requested):
Just a few comments/suggestions but overall it looks good to me!

sergioeliot2039 commented:
@jcortes this is ready for a new review.
as per discussed [on Slack in this Thread](https://pipedream-users.slack.com/archives/G01J50YGJGP/p1645187574309249)
the following review comments were not addressed, since they pertain to Sendgrid sources and we are addressing Sendgrid actions in this PR:
1. [make `axios` lib not called other functions, make it abstracted](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r750626791)

2. [migrating from axios lib to @pipedream/platform](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r750626697) according to this [Pipedream docs](https://pipedream.com/docs/pipedream-axios/#pipedream-platform-axios)

3. [having certain values in constants](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r750626755)

sergioeliot2039 commented:
@jcortes ready for a new review. thanks

jcortes commented on components/sendgrid/actions/common.js:
```diff
@@ -72,10 +72,14 @@ module.exports = {
+          return value || undefined;
+        }
       }
     },
```
How about this
```suggestion
    convertEmptyStringToUndefined(value) {
      if (value === "" || value === null) {
        return undefined;
      }
      return value ?? undefined;
    },
```

jcortes reviewed (changes requested):
Hi @sergioeliot2039 I saw a few suggestions that are still pending to be addressed. Other than that I've just added a new suggestion for the utils function `convertEmptyStringToUndefined` and that would be it. Thanks!

https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819804876
https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819804423
https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819805894

sergioeliot2039 commented:
> Hi @sergioeliot2039 I saw a few suggestions that are still pending to be addressed. Other than that I've just added a new suggestion for the utils function `convertEmptyStringToUndefined` and that would be it. Thanks!
> 
> [#1288 (comment)](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819804876) [#1288 (comment)](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819804423) [#1288 (comment)](https://github.com/PipedreamHQ/pipedream/pull/1288#discussion_r819805894)

You are right. I did previously addressed your review comment about using `await` keyword but just actually missed in some instances. Since many methods under the sendgrid app were affected, i iterated a test across all components.

@jcortes this is ready for a new review.

js07 commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,77 @@
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    const listBlocksEndpoint = "/v3/suppression/blocks";
+    return this.sendgrid.listItems(
+      listBlocksEndpoint,
+      this.startTime,
+      this.endTime,
+      this.numberOfBlocks,
+    );
```
To encapsulate knowledge of the API in the [app file](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#good-api-usage), I'd suggest moving the `listBlocksEndpoint` (`/v3/suppression/blocks`) to the app file so it can be reused. For example:

- In `sendgrid.app.js`
```js
    async listBlocks(startTime, endTime, maxItems) {
      const listItemsEndpoint = "/v3/suppression/blocks";
      return this.listItems(listItemsEndpoint, startTime, endTime, maxItems);
    },
```

- In `list-blocks.js`
```js
    this.checkValidationResults(validationResult);
    return this.sendgrid.listBlocks(
      this.startTime,
      this.endTime,
      this.numberOfBlocks,
    );
```

sergioeliot2039 replied:
Nice! This suggestion has been implemented.

js07 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,77 @@
+      },
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    const listSupressionsEndpoint = "/v3/suppression/unsubscribes";
+    return this.sendgrid.listItems(
+      listSupressionsEndpoint,
+      this.startTime,
+      this.endTime,
+      this.numberOfSupressions,
+    );
```
To encapsulate knowledge of the API in the [app file](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#good-api-usage), I'd suggest moving the `listBlocksEndpoint` (`/v3/suppression/unsubscribes`) to the app file so it can be reused. For example:

- In `sendgrid.app.js`
```js
    async listUnsubscribes(startTime, endTime, maxItems) {
      const listItemsEndpoint = "/v3/suppression/unsubscribes";
      return this.listItems(listItemsEndpoint, startTime, endTime, maxItems);
    },
```

- In `list-global-supresions.js`
```js
    this.checkValidationResults(validationResult);
    return this.sendgrid.listUnsubscribes(
      this.startTime,
      this.endTime,
      this.numberOfSuppressions,
    );
```

sergioeliot2039 replied:
Ok, I have implemented this as well, however. I preserved the method name `listGlobalSupressions` to be consistent with the API documentation, where the API reference entry is named "Retrieve all global suppressions " but the related API endpoint is related to unsubscribes:

> Retrieve all global suppressions
> GET /v3/suppression/unsubscribes

See [Retrieve all global suppressions](https://docs.sendgrid.com/api-reference/suppressions-global-suppressions/retrieve-all-global-suppressions) at the Sendgrid's API Referce.

But please let me know if keeping listUnsubscribesis better for user readiability or any other reason.

js07 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,77 @@
+
+module.exports = {
+  ...common,
+  key: "sendgrid-list-global-supressions",
+  name: "List Global Supressions",
```
**-Minor-**
"Suppressions" in the key, name, and folder/filename is missing a "p".

```suggestion
  key: "sendgrid-list-global-suppressions",
  name: "List Global Suppressions",
```

js07 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,77 @@
+    },
+    numberOfSupressions: {
+      type: "integer",
+      label: "Max # of Global Supressions to Return",
+      description: "Indicates the max number of global supressions to return.",
```
```suggestion
      label: "Max # of Global Suppressions to Return",
      description: "Indicates the max number of global suppressions to return.",
```

js07 reviewed (commented):
@sergioeliot2039 I made some comments for minor refactoring, removing some unnecessary code, bumping source versions, and fixing a spelling issue.

sergioeliot2039 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+      ip_pool_name: this.convertEmptyStringToUndefined(this.ipPoolName),
+      mail_settings: this.convertEmptyStringToUndefined(this.mailSettings),
+      tracking_settings: this.convertEmptyStringToUndefined(this.trackingSettings),
+    };
```
Nice improvement to the code base. I have implemented this suggestion. Thanks

sergioeliot2039 commented:
This is ready for a new review @js07 I think @michelle0927 is on call for PR review this week but anyway. Thanks engaging in all these!

michelle0927 commented on components/sendgrid/actions/add-email-to-global-supression/add-email-to-global-supression.js:
```diff
@@ -0,0 +1,23 @@
+module.exports = {
+  ...common,
+  key: "sendgrid-add-email-to-global-supression",
+  name: "Add Email To Global Supression",
```
```suggestion
  name: "Add Email to Global Supression",
```

michelle0927 commented on components/sendgrid/actions/add-email-to-global-supression/add-email-to-global-supression.js:
```diff
@@ -0,0 +1,23 @@
+        "An string array of email addresses to be added to the global suppressions group. Example `[\"email1@example.com\",\"email2@example.com\"]`",
+    },
+  },
+  async run() {
+    return this.sendgrid.addEmailToGlobalSupression(this.recipientEmails);
+  },
```
Need to include the `$` variable and export a summary for all actions.
```suggestion
  async run({ $ }) {
    const resp = await this.sendgrid.addEmailToGlobalSupression(this.recipientEmails);
    $.export("$summary", "Successfully added emails to global supressions group");
    return resp;
  },
```

michelle0927 commented on components/sendgrid/actions/add-or-update-contact/add-or-update-contact.js:
```diff
@@ -0,0 +1,138 @@
+module.exports = {
+  ...common,
+  key: "sendgrid-add-or-update-contact",
+  name: "Add Or Update Contact",
```
```suggestion
  name: "Add or Update Contact",
```

michelle0927 commented on components/sendgrid/actions/add-or-update-contact/add-or-update-contact.js:
```diff
@@ -0,0 +1,138 @@
+  ...common,
+  key: "sendgrid-add-or-update-contact",
+  name: "Add Or Update Contact",
+  description: "Adds or updates a contact.",
```
All actions need to include a link to the relevant documentation in the description just like you have in `add-email-to-global-supression.js`

michelle0927 commented on components/sendgrid/actions/add-or-update-contact/add-or-update-contact.js:
```diff
@@ -0,0 +1,138 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/delete-contacts/delete-contacts.js:
```diff
@@ -0,0 +1,36 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed
```suggestion
```

michelle0927 commented on components/sendgrid/actions/delete-global-supression/delete-global-supression.js:
```diff
@@ -0,0 +1,36 @@
+  ...common,
+  key: "sendgrid-delete-global-supression",
+  name: "Delete Global Supression",
+  description:
+    "Allows you to remove an email address from the global suppressions group.",
```
Need to add link to relevant documentation.

michelle0927 commented on components/sendgrid/actions/delete-global-supression/delete-global-supression.js:
```diff
@@ -0,0 +1,36 @@
+        "The email address you want to remove from the global suppressions group.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/delete-global-supression/delete-global-supression.js:
```diff
@@ -0,0 +1,36 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/delete-global-supression/delete-global-supression.js:
```diff
@@ -0,0 +1,36 @@
+      email: this.email,
+    }, constraints);
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.deleteGlobalSupression(this.email);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.deleteGlobalSupression(this.email);
    $.export("$summary", `Successfully removed ${this.email} from global supression group.`);
    return resp;
```

michelle0927 commented on components/sendgrid/actions/delete-list/delete-list.js:
```diff
@@ -0,0 +1,31 @@
+      default: false,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/delete-list/delete-list.js:
```diff
@@ -0,0 +1,31 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/delete-list/delete-list.js:
```diff
@@ -0,0 +1,31 @@
+  },
+  async run() {
+    this.deleteContacts = !!this.deleteContacts;
+    return this.sendgrid.deleteList(this.id, this.deleteContacts);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.deleteList(this.id, this.deleteContacts);
    $.export("$summary", `Successfully deleted list ${this.list}.`);
    return resp;
```

michelle0927 commented on components/sendgrid/actions/get-a-block/get-a-block.js:
```diff
@@ -0,0 +1,34 @@
+  ...common,
+  key: "sendgrid-get-a-block",
+  name: "Get a Block",
+  description: "Gets a specific block.",
```
Need to add link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/get-a-block/get-a-block.js:
```diff
@@ -0,0 +1,34 @@
+      description: "The email address of the specific block.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/get-a-block/get-a-block.js:
```diff
@@ -0,0 +1,34 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/get-a-block/get-a-block.js:
```diff
@@ -0,0 +1,34 @@
+      email: this.email,
+    }, constraints);
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.getBlock(this.email);
```
Export a summary.
```suggestion
    const resp = this.sendgrid.getBlock(this.email);
    $.export("$summary", "Successfully retrieved block.");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/get-a-global-supression/get-a-global-supression.js:
```diff
@@ -0,0 +1,35 @@
+  ...common,
+  key: "sendgrid-get-a-global-supression",
+  name: "Get A Global Supression",
+  description: "Gets a global supression.",
```
Need to add a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/get-a-global-supression/get-a-global-supression.js:
```diff
@@ -0,0 +1,35 @@
+        "The email address of the global suppression you want to retrieve.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed
```suggestion
```

michelle0927 commented on components/sendgrid/actions/get-a-global-supression/get-a-global-supression.js:
```diff
@@ -0,0 +1,35 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/get-a-global-supression/get-a-global-supression.js:
```diff
@@ -0,0 +1,35 @@
+      email: this.email,
+    }, constraints);
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.getGlobalSupression(this.email);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.getGlobalSupression(this.email);
    $.export("$summary", "Successfully retrieved global supression.");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/get-all-bounces/get-all-bounces.js:
```diff
@@ -0,0 +1,59 @@
+  ...common,
+  key: "sendgrid-get-all-bounces",
+  name: "Get All Bounces",
+  description: "Allows you to get all of your bounces.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/get-all-bounces/get-all-bounces.js:
```diff
@@ -0,0 +1,59 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed
```suggestion
```

michelle0927 commented on components/sendgrid/actions/get-all-bounces/get-all-bounces.js:
```diff
@@ -0,0 +1,59 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/get-all-bounces/get-all-bounces.js:
```diff
@@ -0,0 +1,59 @@
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.getAllBounces(this.startTime, this.endTime);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.getAllBounces(this.startTime, this.endTime);
    $.export("$summary", "Successfully retrieved bounces.");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/get-contact-lists/get-contact-lists.js:
```diff
@@ -0,0 +1,41 @@
+  ...common,
+  key: "sendgrid-get-contact-lists",
+  name: "Get Contact Lists",
+  description: "Allows you to get details of your contact lists.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/get-contact-lists/get-contact-lists.js:
```diff
@@ -0,0 +1,41 @@
+      description: "The number of contact lists to return.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/get-contact-lists/get-contact-lists.js:
```diff
@@ -0,0 +1,41 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/get-contact-lists/get-contact-lists.js:
```diff
@@ -0,0 +1,41 @@
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.getAllContactLists(this.numberOfLists);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.getAllContactLists(this.numberOfLists);
    $.export("$summary", "Successfully retrieved lists");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,75 @@
+  ...common,
+  key: "sendgrid-list-blocks",
+  name: "List Blocks",
+  description:
+    "Allows you to list all email addresses that are currently on your blocks list.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,75 @@
+      description: "Indicates the max number of blocked emails to return.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,75 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/list-blocks/list-blocks.js:
```diff
@@ -0,0 +1,75 @@
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.listBlocks(
+      this.startTime,
+      this.endTime,
+      this.numberOfBlocks,
+    );
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.listBlocks(
      this.startTime,
      this.endTime,
      this.numberOfBlocks,
    );
    $.export("$summary", "Successsfully retrieved blocks");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,75 @@
+  ...common,
+  key: "sendgrid-list-global-suppressions",
+  name: "List Global Suppressions",
+  description:
+    "Allows you to get a list of all email address that are globally suppressed.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,75 @@
+      description: "Indicates the max number of global suppressions to return.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,75 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/list-global-supressions/list-global-supressions.js:
```diff
@@ -0,0 +1,75 @@
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    return this.sendgrid.listGlobalSupressions(
+      this.startTime,
+      this.endTime,
+      this.numberOfSuppressions,
+    );
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.listGlobalSupressions(
      this.startTime,
      this.endTime,
      this.numberOfSuppressions,
    );
    $.export("$summary", "Successfully retrieved global supressions");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,31 @@
+  ...common,
+  key: "sendgrid-remove-a-contact-from-a-list",
+  name: "Remove A Contact From A List",
+  description: "Allows you to remove contacts from a given list.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,31 @@
+  type: "action",
+  props: {
+    ...common.props,
+    id: {
+      type: "string",
+      label: "Id",
+      description:
+        "Unique Id of the List where the contact to remove off is located.",
+    },
```
Can `id` use `async options`?

michelle0927 commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,31 @@
+      description:
+        "Unique Id of the List where the contact to remove off is located.",
+    },
+    contactIds: {
+      type: "string[]",
+      label: "Contact Ids",
+      description:
+        "String array of contact ids to be removed off the list.",
+    },
```
Could `contactIds` use `async options`?

michelle0927 commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,31 @@
+        "String array of contact ids to be removed off the list.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/remove-a-contact-from-a-list/remove-a-contact-from-a-list.js:
```diff
@@ -0,0 +1,31 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
+    return this.sendgrid.removeContactFromList(this.id, this.contactIds);
+  },
```
Include the `$` variable and export a summary.
```suggestion
  async run({ $ }) {
    const resp = await this.sendgrid.removeContactFromList(this.id, this.contactIds);
    $.export("$summary", `Successfully removed contact ${this.contactId} from list.`);
    return resp;
  },
```

michelle0927 commented on components/sendgrid/actions/search-contacts/search-contacts.js:
```diff
@@ -0,0 +1,25 @@
+  ...common,
+  key: "sendgrid-search-contacts",
+  name: "Search Contacts",
+  description: "Searches contacts with a SGQL query.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/search-contacts/search-contacts.js:
```diff
@@ -0,0 +1,25 @@
+        "The query field accepts valid SGQL for searching for a contact. Only the first 50 contacts will be returned. If the query takes longer than 20 seconds a 408 Request Timeout will be returned.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/search-contacts/search-contacts.js:
```diff
@@ -0,0 +1,25 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
+    return this.sendgrid.searchContacts(this.query);
+  },
```
Include the `$` variable and export a summary.
```suggestion
  async run({ $ }) {
    const resp = await this.sendgrid.searchContacts(this.query);
    $.export("$summary", "Successfully completed search");
    return resp;
  },
```

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+  ...common,
+  key: "sendgrid-send-email-multiple-recipients",
+  name: "Send Email Multiple Recipients",
+  description:
+    "This action sends a personalized e-mail to multiple specified recipients.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+      description:
+        "An array of messages and their metadata. Each object within personalizations can be thought of as an envelope - it defines who should receive an individual message and how that message should be handled. See our [Personalizations documentation](https://sendgrid.com/docs/for-developers/sending-email/personalizations/) for details. maxItems: 1000. Example: `[{to:[{email:\"email@email.com\" ...
+    },
+    fromEmail: {
+      type: "string",
+      label: "From Email",
+      description:
+        "The 'From' email address used to deliver the message. This address should be a verified sender in your Twilio SendGrid account.",
+    },
```
Could `fromEmail` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+      "An array of objects where you can specify any attachments you want to include. The fields `content` and `filename` are required. `content` must be base64 encoded. Alternatively, provide a string that will `JSON.parse` to an array of attachments objects. Example: `[{content:\"aGV5\",type:\"text/plain\",filename:\"sample.txt\"}]`",
+      optional: true,
+    },
+    templateId: {
+      type: "string",
+      label: "Template Id",
+      description:
+        "An email template ID. A template that contains a subject and content — either text or html — will override any subject and content values specified at the personalizations or message level.",
+      optional: true,
+    },
```
Could `templateId` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+        "A unix timestamp allowing you to specify when you want your email to be delivered. This may be overridden by the `send_at` parameter set at the personalizations level. Delivery cannot be scheduled more than 72 hours in advance. If you have the flexibility, it's better to schedule mail for off-peak times. Most emails are scheduled and sent at the top of the hour or half hour. Scheduling e ...
+      optional: true,
+    },
+    batchId: {
+      type: "integer",
+      label: "Batch Id",
+      description:
+        "An ID representing a batch of emails to be sent at the same time. Including a batch_id in your request allows you include this email in that batch. It also enables you to cancel or pause the delivery of that batch. For more information, see the [Cancel Scheduled Sends API](https://sendgrid.com/docs/api-reference/).",
+      optional: true,
+    },
```
Could `batchId` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/send-email-multiple-recipients/send-email-multiple-recipients.js:
```diff
@@ -0,0 +1,242 @@
+      mail_settings: this.mailSettings,
+      tracking_settings: this.trackingSettings,
+    });
+    return this.sendgrid.sendEmail(config);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.sendEmail(config);
    $.export("$summary", "Email successfully sent");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+  ...common,
+  key: "sendgrid-send-email-single-recipient",
+  name: "Send Email Single Recipient",
+  description:
+    "This action sends a personalized e-mail to the specified recipient.",
```
Need to add a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+  type: "action",
+  props: {
+    ...common.props,
+    fromEmail: {
+      type: "string",
+      label: "From Email",
+      description:
+        "The 'From' email address used to deliver the message. This address should be a verified sender in your Twilio SendGrid account.",
+    },
```
Could `fromEmail` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+      "An array of objects where you can specify any attachments you want to include. The fields `content` and `filename` are required. `content` must be base64 encoded. Alternatively, provide a string that will `JSON.parse` to an array of attachments objects. Example: `[{content:\"aGV5\",type:\"text/plain\",filename:\"sample.txt\"}]`",
+      optional: true,
+    },
+    templateId: {
+      type: "string",
+      label: "Template Id",
+      description:
+        "An email template ID. A template that contains a subject and content — either text or html — will override any subject and content values specified at the personalizations or message level.",
+      optional: true,
+    },
```
Could `templateId` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+        "A unix timestamp allowing you to specify when you want your email to be delivered. This may be overridden by the `send_at` parameter set at the personalizations level. Delivery cannot be scheduled more than 72 hours in advance. If you have the flexibility, it's better to schedule mail for off-peak times. Most emails are scheduled and sent at the top of the hour or half hour. Scheduling e ...
+      optional: true,
+    },
+    batchId: {
+      type: "integer",
+      label: "Batch Id",
+      description:
+        "An ID representing a batch of emails to be sent at the same time. Including a batch_id in your request allows you include this email in that batch. It also enables you to cancel or pause the delivery of that batch. For more information, see the [Cancel Scheduled Sends API](https://sendgrid.com/docs/api-reference/).",
+      optional: true,
+    },
```
Could `batchId` use `async options`?

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/send-email-single-recipient/send-email-single-recipient.js:
```diff
@@ -0,0 +1,325 @@
+      mail_settings: this.mailSettings,
+      tracking_settings: this.trackingSettings,
+    });
+    return this.sendgrid.sendEmail(config);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.sendEmail(config);
    $.export("$summary", "Email successfully sent");
    return resp;
```

michelle0927 commented on components/sendgrid/actions/validate-email/validate-email.js:
```diff
@@ -0,0 +1,49 @@
+  ...common,
+  key: "sendgrid-validate-email",
+  name: "Validate Email",
+  description:
+    "Validates an email address. This action requires a Sendgrid's Pro or Premier plan.",
```
Need to include a link to the relevant documentation.

michelle0927 commented on components/sendgrid/actions/validate-email/validate-email.js:
```diff
@@ -0,0 +1,49 @@
+      optional: true,
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
```
This can be removed.
```suggestion
```

michelle0927 commented on components/sendgrid/actions/validate-email/validate-email.js:
```diff
@@ -0,0 +1,49 @@
+  methods: {
+    ...common.methods,
+  },
+  async run() {
```
Include the `$` variable.
```suggestion
  async run({ $ }) {
```

michelle0927 commented on components/sendgrid/actions/validate-email/validate-email.js:
```diff
@@ -0,0 +1,49 @@
+      email: this.email,
+      source: this.source,
+    };
+    return this.sendgrid.validateEmail(body);
```
Export a summary.
```suggestion
    const resp = await this.sendgrid.validateEmail(body);
    $.export("$summary", "Email validation completed");
    return resp;
```

michelle0927 commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -1,6 +1,7 @@
 const axios = require("axios");
```
Could this be updated to use Pipedream's axios? https://pipedream.com/docs/v1/pipedream-axios/#pipedream-platform-axios
```suggestion
const { axios } = require("@pipedream/platform");
```
or in ES6
```suggestion
import { axios } from "@pipedream/platform"
```

michelle0927 commented on components/sendgrid/sendgrid.app.js:
```diff
@@ -12,9 +13,17 @@ module.exports = {
       return "https://api.sendgrid.com/v3";
     },
-    _contactsSearchUrl() {
+    api() {
+      sendgrid.setDefaultHeader({
+        "Content-Type": "application/json",
+        "User-Agent": "@PipedreamHQ/pipedream v0.1",
+      });
+      sendgrid.setApiKey(this.$auth.api_key);
+      return sendgrid;
+    },
```
Might as well use `_authToken()` for getting `this.$auth.api_key`.
```suggestion
    api() {
      sendgrid.setDefaultHeader({
        "Content-Type": "application/json",
        "User-Agent": "@PipedreamHQ/pipedream v0.1",
      });
      sendgrid.setApiKey(this._authToken());
      return sendgrid;
    },
```

michelle0927 reviewed (changes requested):
Do you think you could update these components to .mjs (ES Modules) instead of .js? 

Btw, I made some comments about using `async options` for some props, but haven't checked the documentation to see if it's possible.

FYI, if you update the app file to use Pipedream's `axios`, you'll need to update the methods that use axios as well. Pipedream's `axios` uses the `$` variable from actions to provide better debugging info. It returns the `data` object from the response.
`axios($ ?? this, config);`
https://pipedream.com/docs/v1/pipedream-axios/#why-pipedream-platform-axios
